File: 20250518-AHC047/02.cpp
#include <bits/stdc++.h>
using namespace std;

constexpr int TIME_LIMIT_MS = 1950;  // 1.9秒
constexpr int SAMPLE_LENGTH = 900;
constexpr int SAMPLES = 10;
constexpr int BEAM_WIDTH = 30;
constexpr int NEIGHBORS_PER_BEAM = 3;
const vector<char> CHARS = {'a','b','c','d','e','f'};

int N, M, L;
vector<pair<string,int>> patterns;

random_device rd;
mt19937 rng(rd());
uniform_int_distribution<int> dist_char(0, (int)CHARS.size() - 1);
uniform_int_distribution<int> dist_prob(0, 99);
uniform_real_distribution<double> dist_real(0.0, 1.0);
uniform_int_distribution<int> dist_delta(-10, 10);

// Cをランダム生成
vector<char> random_C() {
    vector<char> C(M);
    for (int i = 0; i < M; i++) {
        C[i] = CHARS[dist_char(rng)];
    }
    return C;
}

// Aをランダム生成（各行100点満点に正規化）
vector<vector<int>> random_A() {
    vector<vector<int>> A(M, vector<int>(M));
    for (int i = 0; i < M; i++) {
        vector<int> prob(M);
        int s = 0;
        for (int j = 0; j < M; j++) {
            prob[j] = uniform_int_distribution<int>(0, 100)(rng);
            s += prob[j];
        }
        for (int j = 0; j < M; j++) {
            A[i][j] = prob[j] * 100 / s;
        }
        // 調整
        int sum_row = accumulate(A[i].begin(), A[i].end(), 0);
        A[i][M-1] += 100 - sum_row;
    }
    return A;
}

// C,Aから長さlengthの文字列を生成
string generate_string(const vector<char> &C, const vector<vector<int>> &A, int length) {
    string s;
    s += C[0];
    int cur = 0;
    for (int i = 1; i < length; i++) {
        int r = dist_prob(rng);
        int acc = 0;
        for (int j = 0; j < M; j++) {
            acc += A[cur][j];
            if (r < acc) {
                cur = j;
                break;
            }
        }
        s += C[cur];
    }
    return s;
}

// 生成文字列中にパターンがあるかチェック
bool contains(const string &s, const string &pat) {
    return s.find(pat) != string::npos;
}

// C,Aの評価：SAMPLES回生成してパターン検出率でスコア計算
int evaluate(const vector<char> &C, const vector<vector<int>> &A) {
    vector<int> found(N, 0);
    for (int i = 0; i < SAMPLES; i++) {
        string s = generate_string(C, A, SAMPLE_LENGTH);
        for (int j = 0; j < N; j++) {
            if (contains(s, patterns[j].first)) found[j]++;
        }
    }
    double score = 0;
    for (int i = 0; i < N; i++) {
        score += (double)found[i] / SAMPLES * patterns[i].second;
    }
    return (int)(score + 0.5);
}

// 近傍生成
pair<vector<char>, vector<vector<int>>> neighbor(const vector<char> &C, const vector<vector<int>> &A) {
    vector<char> newC = C;
    vector<vector<int>> newA = A;

    if (dist_real(rng) < 0.5) {
        // Cを1文字ランダムに変更
        int i = uniform_int_distribution<int>(0, M-1)(rng);
        newC[i] = CHARS[dist_char(rng)];
    } else {
        // Aの1要素を±delta変更
        int i = uniform_int_distribution<int>(0, M-1)(rng);
        int j = uniform_int_distribution<int>(0, M-1)(rng);
        int delta = dist_delta(rng);
        int val = newA[i][j] + delta;
        if (0 <= val && val <= 100) {
            newA[i][j] = val;
            // 再正規化
            int total = accumulate(newA[i].begin(), newA[i].end(), 0);
            for (int k = 0; k < M; k++) {
                newA[i][k] = newA[i][k] * 100 / total;
            }
            newA[i][M-1] += 100 - accumulate(newA[i].begin(), newA[i].end(), 0);
        }
    }
    return {newC, newA};
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M >> L;
    patterns.resize(N);
    for (int i = 0; i < N; i++) {
        string s; int p;
        cin >> s >> p;
        patterns[i] = {s, p};
    }

    // 初期化
    auto bestC = random_C();
    auto bestA = random_A();
    int bestScore = evaluate(bestC, bestA);

    using State = tuple<int, vector<char>, vector<vector<int>>>; // score, C, A
    vector<State> beam;
    beam.emplace_back(bestScore, bestC, bestA);

    auto start = chrono::steady_clock::now();

    while (true) {
        auto now = chrono::steady_clock::now();
        auto elapsed_ms = chrono::duration_cast<chrono::milliseconds>(now - start).count();
        if (elapsed_ms > TIME_LIMIT_MS) break;

        vector<State> new_beam;
        for (auto &[score, C, A] : beam) {
            for (int _ = 0; _ < NEIGHBORS_PER_BEAM; _++) {
                auto [C2, A2] = neighbor(C, A);
                int sc = evaluate(C2, A2);
                new_beam.emplace_back(sc, C2, A2);
            }
        }
        sort(new_beam.rbegin(), new_beam.rend());
        if (new_beam.size() > BEAM_WIDTH) new_beam.resize(BEAM_WIDTH);

        beam = move(new_beam);
        if (get<0>(beam[0]) > bestScore) {
            bestScore = get<0>(beam[0]);
            bestC = get<1>(beam[0]);
            bestA = get<2>(beam[0]);
        }
    }

    // 出力
    for (int i = 0; i < M; i++) {
        cout << bestC[i];
        for (int j = 0; j < M; j++) {
            cout << " " << bestA[i][j];
        }
        cout << "\n";
    }

    return 0;
}

File: AHC/20250518-AHC047/02.py
import sys
import time
import random
from collections import defaultdict

# --- 入力の読み込み ---
N, M, L = map(int, input().split())
patterns = [input().split() for _ in range(N)]
patterns = [(s, int(p)) for s, p in patterns]

CHARS = list('abcdef')
START_TIME = time.time()
TIME_LIMIT = 1.9  # 秒以内に出力する（マージン込み）
SAMPLES = 20
SAMPLE_LENGTH = 1000
BEAM_WIDTH = 10

# --- ユーティリティ関数群 ---

def random_C():
    return [random.choice(CHARS) for _ in range(M)]

def random_A():
    A = []
    for _ in range(M):
        prob = [random.randint(0, 100) for _ in range(M)]
        s = sum(prob)
        norm = [x * 100 // s for x in prob]
        norm[-1] += 100 - sum(norm)  # 合計が100になるよう調整
        A.append(norm)
    return A

def generate_string(C, A, length):
    s = [C[0]]
    cur = 0
    for _ in range(length - 1):
        r = random.randint(0, 99)
        acc = 0
        for j in range(M):
            acc += A[cur][j]
            if r < acc:
                cur = j
                break
        s.append(C[cur])
    return ''.join(s)

def evaluate(C, A, patterns, samples=SAMPLES):
    found = [0] * len(patterns)
    for _ in range(samples):
        s = generate_string(C, A, SAMPLE_LENGTH)
        for i, (pat, _) in enumerate(patterns):
            if pat in s:
                found[i] += 1
    return round(sum((found[i] / samples) * p for i, (_, p) in enumerate(patterns)))

def neighbor(C, A):
    newC = C[:]
    newA = [row[:] for row in A]

    if random.random() < 0.5:
        i = random.randint(0, M - 1)
        newC[i] = random.choice(CHARS)
    else:
        i = random.randint(0, M - 1)
        j = random.randint(0, M - 1)
        delta = random.randint(-10, 10)
        if 0 <= newA[i][j] + delta <= 100:
            newA[i][j] += delta
            total = sum(newA[i])
            if total != 100:
                # 再正規化
                for k in range(M):
                    newA[i][k] = newA[i][k] * 100 // total
                newA[i][-1] += 100 - sum(newA[i])
    return newC, newA

# --- 初期化 ---
bestC = random_C()
bestA = random_A()
bestScore = evaluate(bestC, bestA, patterns)

# --- 焼きなまし + ビームサーチ ---
beam = [(bestScore, bestC, bestA)]

while time.time() - START_TIME < TIME_LIMIT:
    new_beam = []
    for score, C, A in beam:
        for _ in range(3):  # 各ビームから3近傍
            C2, A2 = neighbor(C, A)
            sc = evaluate(C2, A2, patterns)
            new_beam.append((sc, C2, A2))
    beam = sorted(new_beam, reverse=True)[:BEAM_WIDTH]
    if beam[0][0] > bestScore:
        bestScore, bestC, bestA = beam[0]

# --- 出力 ---
for i in range(M):
    row = [str(bestA[i][j]) for j in range(M)]
    print(f"{bestC[i]} {' '.join(row)}")
